data-processing/common/parse_equation.py

```diff
@@ -1,151 +1,290 @@
-                consecutive_char_sequence = []
-                consecutive_char_indices = []
-            # Clean the MathML of our annotations. Do this cleaning on a clone of the node, so that the
-            # annotations are available for processing later nodes.
-            node_clone = _clean_node_of_annotations(symbol_node)
-            node_symbols.append(
-                # set node to 'symbol_node', not clone, as the node must have a parent for the
-                # 'set_children' method. The clone will have had its parent removed.
-                NodeSymbol(characters=characters, mathml=str(node_clone), node=symbol_node)
+            soup_children = merged_children
+
+    # Extract the tokens that are uniquely defined in this element.
+    tokens = _extract_tokens(element)
+
+    # Create a new BeautifulSoup element that will be a cleaned clone of the input element.
+    # Remove children from the element. Cleaned up children will be added shortly.
+    clean_element = copy.copy(element)
+    _remove_s2_annotations(clean_element)
+    clean_element.clear()
+
+    # Visit each of the child nodes. Iterate over 'element' instead of 'clean_element', as the
+    # children will have been removed from 'clean_element' As of BeautifulSoup version 4.8.2,
+    # children are visited in order.
+    child_symbols = []
+    for child in soup_children:
+        if isinstance(child, Tag):
+            child_parse_result = parse_element(child)
+            child_symbols.extend(child_parse_result.symbols)
+            if child_parse_result.element is not None:
+                clean_element.append(child_parse_result.element)
+
+            # Register the tokens found in the child elements as being part of this symbol.
+            tokens.extend(child_parse_result.tokens)
+
+        # If the child is not a tag, it's probably a string (like 'x' in <mi>x</mi>). We need
+        # to make sure these are getting added to the element_clone too.
+        elif isinstance(child, str) and not child.isspace():
+            clean_element.append(child)
+
+    # If the element's tag indicates the element is a symbol, return a new symbol node, setting its
+    # children to the symbols found in the children.
+    if element.name in SYMBOL_TAGS:
+        return_symbols = [
+            Node(children=child_symbols, element=clean_element, tokens=tokens)
+        ]
+    # Otherwise, return the list of symbols found in the children.
+    else:
+        return_symbols = child_symbols
+
+    return ParseResult(return_symbols, clean_element, tokens)
+
+
+def _extract_tokens(element: Tag) -> List[Token]:
+    """
+    Get the tokens defined in this element. Tokens are only found in low-level elements like
+    "<mi>" and "<mn>". This function will find no tokens in higher-level nodes that solely
+    group other low-level elements (like "<mrow>" and "<msub>").
+    """
+
+    tokens = []
+    if element.name in ["mi", "mn"] and _has_s2_token_annotations(element):
```

Comment:
curious why `["mi", "mn"]` is not a defined constant like SYMBOL_TAGS?

Reply:
Currently because it's used only in one place, though do you think readability would be improved with a defined constant?

Reply:
Ah that makes sense,  is there a connection between this ["mi", "mn"] and the one in `_is_mergeable_type` or is it pure coincidence? If not, it might make sense to define then

Reply:
Good point, they're slightly semantically different, and I think in the future they might diverge. The first list is the set of all tokens; the second is the list of all tokens that should be considered as mergeable with each other (which, in the future, may not contain all tokens, e.g., operators). I gave them names! 👍